main.py

```diff
@@ -35,26 +42,36 @@
 # [START run_pubsub_handler]
 @app.route('/', methods=['POST'])
 def index():
-    envelope = request.get_json()
-    if not envelope:
-        msg = 'no Pub/Sub message received'
-        print(f'error: {msg}')
-        return f'Bad Request: {msg}', 400
+    pubsub_received_message = request.get_json()
+
+    # parse the Pub/Sub data
+    try:
+        pubsub_data_string = pubsub.parse_data_from_message(pubsub_received_message)
+    except pubsub.DataParseError as e:
+        print(e)
+        return (e.message, 400)
+
+    # load the notification from the data
+    try:
+        monitoring_notification_dict = json.loads(pubsub_data_string)
+    except json.JSONDecodeError as e:
+        msg = 'notification could not be decoded to a JSON'
+        print(msg)
+        return (msg, 400)
 
-    if not isinstance(envelope, dict) or 'message' not in envelope:
-        msg = 'invalid Pub/Sub message format'
-        print(f'error: {msg}')
-        return f'Bad Request: {msg}', 400
 
-    pubsub_message = envelope['message']
+    philips_hue_client = philips_hue.PhilipsHueClient(app.config['BRIDGE_IP_ADDRESS'],
+                                                      app.config['USERNAME'])
 
-    name = 'World'
-    if isinstance(pubsub_message, dict) and 'data' in pubsub_message:
-        name = base64.b64decode(pubsub_message['data']).decode('utf-8').strip()
+    try:
+        response = philips_hue.trigger_light_from_monitoring_notification(
+            philips_hue_client, monitoring_notification_dict, app.config['LIGHT_ID'])
+    except philips_hue.Error as e:
+        print(e)
+        return (e.message, 400)
 
-    print(f'Hello {name}!')
 
-    return ('', 204)
+    return (response.text, 200)
```

Comment:
Just checking: do we have to do anything to ack the pubsub message, other than return `200`?

Also, just pointing out, I'm not sure that the correct behavior is to not-ack messages that encounter errors. For instance, if the message is missing a required dict key and raises an exception, so then you don't ack the message, PubSub will try to deliver to another subscriber. But then that subscriber will run into the same exception, refuse to ack, trigger a redelivery, and so on.

At least that's my understanding. I would check to make sure that I'm correct.

If my mental model is correct, I think that we want to ack messages that encounter permanent errors, that will fail even if we retry.

Reply:
Based on the Pub/Sub docs (https://cloud.google.com/pubsub/docs/subscriber) I think an acknowledgement is indicated just with an HTTP success status code. This is what the docs say:

"The endpoint acknowledges the message by returning an HTTP success status code. A non-success response indicates that the message should be resent."

So it does look like if there is an error the message is redelivered. I'm not super sure how to get around this; the sample repo that we based this project off of returned an error code when the pubsub message format was incorrect as well. Since I haven't worked with Pub/Sub in the past I don't know what the best practice is, but if an error isn't addressed maybe it's ok that the message is delivered multiple times? The error will still be logged in Cloud Run and it might be good to resend the error if it isn't being fixed/addressed within the ackDeadline. Another thing is if the ackDeadline is set to be longer, Pub/Sub won't resend a message until the ackDeadline has passed, which should help with reducing the redelivery frequency. 

But if the purpose of acknowledging is just to indicate that a response from pub/sub was received, then maybe we should just return status code 200 with the error message so that in the logs the client can still see that there was a problem, although having all the responses return status code 200 might not be the most helpful.

I don't really know what the best answer to this is, but these are some of my thoughts.

Reply:
I see, thanks for doing some research. it sounds like there's some differences between internal and cloud pubsub (or, maybe, I misconfigured things internally). I would follow what the docs demonstrate, in that case, and return failure codes.